{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1125c161-6952-44ee-9d6b-af04aa35d04d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NZ_CP139575.1\n",
      "Nome: NZ_CP139575\n",
      "Descrição: Lactobacillus acidophilus strain ATCC 4356 chromosome, complete genome\n",
      "Anotações:\n",
      "  molecule_type: DNA\n",
      "  topology: circular\n",
      "  data_file_division: CON\n",
      "  date: 09-DEC-2024\n",
      "  accessions: ['NZ_CP139575']\n",
      "  sequence_version: 1\n",
      "  keywords: ['RefSeq']\n",
      "  source: Lactobacillus acidophilus\n",
      "  organism: Lactobacillus acidophilus\n",
      "  taxonomy: ['Bacteria', 'Bacillati', 'Bacillota', 'Bacilli', 'Lactobacillales', 'Lactobacillaceae', 'Lactobacillus']\n",
      "  references: [Reference(title='The Complete Genome Sequence of Probiotic Lactobacillus acidophilus ATCC 9224 isolated from sour milk', ...), Reference(title='Direct Submission', ...)]\n",
      "  comment: REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CP139575.1.\n",
      "Bacteria and source DNA available from Anand Kumar Bikini Atoll Rd,\n",
      "SM-30, Los Alamos National Lab, Los Alamos, NM.\n",
      "The annotation was added by the NCBI Prokaryotic Genome Annotation\n",
      "Pipeline (PGAP). Information about PGAP can be found here:\n",
      "https://www.ncbi.nlm.nih.gov/genome/annotation_prok/\n",
      "COMPLETENESS: full length.\n",
      "  structured_comment: defaultdict(<class 'dict'>, {'Genome-Assembly-Data': {'Assembly Date': '17-JUL-2017', 'Assembly Method': 'HGAP v. 3.0', 'Genome Representation': 'Full', 'Expected Final Version': 'Yes', 'Genome Coverage': '25.0x', 'Sequencing Technology': 'PacBio RSII'}, 'Genome-Annotation-Data': {'Annotation Provider': 'NCBI RefSeq', 'Annotation Name': 'GCF_034298135.1-RS_2024_12_09', 'Annotation Date': '12/09/2024 18:47:07', 'Annotation Pipeline': 'NCBI Prokaryotic Genome Annotation Pipeline (PGAP)', 'Annotation Method': 'Best-placed reference protein set; GeneMarkS-2+', 'Annotation Software revision': '6.9', 'Features Annotated': 'Gene; CDS; rRNA; tRNA; ncRNA', 'Genes (total)': '1,917', 'CDSs (total)': '1,842', 'Genes (coding)': '1,763', 'CDSs (with protein)': '1,763', 'Genes (RNA)': '75', 'rRNAs': '4, 4, 4 (5S, 16S, 23S)', 'complete rRNAs': '4, 4, 4 (5S, 16S, 23S)', 'tRNAs': '60', 'ncRNAs': '3', 'Pseudo Genes (total)': '79', 'CDSs (without protein)': '79', 'Pseudo Genes (ambiguous residues)': '0 of 79', 'Pseudo Genes (frameshifted)': '59 of 79', 'Pseudo Genes (incomplete)': '17 of 79', 'Pseudo Genes (internal stop)': '16 of 79', 'Pseudo Genes (multiple problems)': '12 of 79', 'CRISPR Arrays': '1'}})\n",
      "  contig: join(CP139575.1:1..1978847)\n",
      "\n",
      "Referências externas:\n",
      "  BioProject:PRJNA224116\n",
      "  BioSample:SAMN38517762\n",
      "  Assembly:GCF_034298135.1\n",
      "----------------------------------------\n",
      "ID: NZ_CP139575.1\n",
      "Tipo de Feature: source\n",
      "Localização: [0:1978847](+)\n",
      "  organism: ['Lactobacillus acidophilus']\n",
      "  mol_type: ['genomic DNA']\n",
      "  strain: ['ATCC 4356']\n",
      "  host: ['Homo sapiens']\n",
      "  type_material: ['type strain of Lactobacillus acidophilus']\n",
      "  db_xref: ['taxon:1579']\n",
      "  geo_loc_name: ['USA']\n",
      "  collection_date: ['2017-07-17']\n",
      "  collected_by: ['ATCC']\n",
      "----------------------------------------\n",
      "Base de Dados: taxon\n",
      "  ID: 1579\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "Entrez.email = \"pg45861@uminho.pt\"\n",
    "\n",
    "def search_sequences(gene_name, db=\"nucleotide\", retmax=10):\n",
    "    handle = Entrez.esearch(db=db, term=gene_name, retmax=retmax)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    ids = record[\"IdList\"]\n",
    "\n",
    "    handle = Entrez.efetch(db=db, id=ids, rettype=\"gb\", retmode=\"text\")\n",
    "    seq_records = list(SeqIO.parse(handle, \"genbank\"))\n",
    "    handle.close()\n",
    "    return seq_records\n",
    "\n",
    "gene_name = \"SO785_RS05220\"\n",
    "seq_records = search_sequences(gene_name)\n",
    "\n",
    "with open(f\"{gene_name}_sequences.gb\", \"w\") as output_file:\n",
    "    SeqIO.write(seq_records, output_file, \"genbank\")\n",
    "\n",
    "def show_annotations(seq_records):\n",
    "    for record in seq_records:\n",
    "        print(f\"ID: {record.id}\")\n",
    "        print(f\"Nome: {record.name}\")\n",
    "        print(f\"Descrição: {record.description}\")\n",
    "        print(\"Anotações:\")\n",
    "        for key, value in record.annotations.items():\n",
    "            print(f\"  {key}: {value}\")\n",
    "        print(\"\\nReferências externas:\")\n",
    "        for dbxref in record.dbxrefs:\n",
    "            print(f\"  {dbxref}\")\n",
    "        print(\"-\" * 40)\n",
    "\n",
    "show_annotations(seq_records)\n",
    "\n",
    "def analize_features(seq_records):\n",
    "    for record in seq_records:\n",
    "        print(f\"ID: {record.id}\")\n",
    "        for feature in record.features:\n",
    "            print(f\"Tipo de Feature: {feature.type}\")\n",
    "            print(f\"Localização: {feature.location}\")\n",
    "            if feature.qualifiers:\n",
    "                for key, value in feature.qualifiers.items():\n",
    "                    print(f\"  {key}: {value}\")\n",
    "        print(\"-\" * 40)\n",
    "\n",
    "analize_features(seq_records)\n",
    "\n",
    "def external_references(seq_records):\n",
    "    references = {}\n",
    "    for record in seq_records:\n",
    "        for feature in record.features:\n",
    "            if \"db_xref\" in feature.qualifiers:\n",
    "                for ref in feature.qualifiers[\"db_xref\"]:\n",
    "                    db, id = ref.split(\":\")\n",
    "                    if db not in references:\n",
    "                        references[db] = []\n",
    "                    references[db].append(id)\n",
    "    return references\n",
    "\n",
    "references = external_references(seq_records)\n",
    "for db, ids in referencias.items():\n",
    "    print(f\"Base de Dados: {db}\")\n",
    "    for id in ids:\n",
    "        print(f\"  ID: {id}\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
